### PR TITLE
RSTcheckBear: Allow ignoring directives and roles

### PR DESCRIPTION
**Pull request by umeshksingla**

### CODE REVIEW

**Comment by umeshksingla** [comment]
WIP.

**Comment by umeshksingla** [comment]
CircleCI fails due to other Bears and not this one, I suppose. Logs say, 

| Name | Stmts | Miss | Branch | BrPart | Cover | Missing |
| --- | --- | --- | --- | --- | --- | --- |
| bears/rest/RSTcheckBear.py  | 24 | 0 | 12 | 0 | 100% | |
| bears/rest/__init__.py  | 0 | 0 | 0 | 0 | 100% | |
| bears/rest/reSTLintBear.py | 18 | 0 | 2 | 0 | 100% | |

**Comment by umeshksingla** [comment]
@Makman2 please review. Thanks.

File: bears/rest/RSTcheckBear.py [hunk 1]
```diff
@@ -32,16 +33,32 @@ class RSTcheckBear:
...
         :param code_block_language_ignore:
             Comma seperated list for languages of which code blocks
             should be ignored. Code block of following languages
             can be detected: ``bash``, ``c``, ``cpp``, ``json``,
             ``python``, ``rst``.
+
+        :param directive_ignore:
+             Comma separated list of directives to be ignored. Can be
+             used to ignore custom directives as well.
+
```
Comment (Makman2) [inline bears/rest/RSTcheckBear.py]: please remove those newlines to conform to the other docstrings in our codebase :)

File: bears/rest/RSTcheckBear.py [hunk 2]
```diff
@@ -32,16 +33,32 @@ class RSTcheckBear:
 
     @staticmethod
     def create_arguments(filename, file, config_file,
-                         code_block_language_ignore: list=()):
+                         code_block_language_ignore: typed_list(str)=(),
```
Comment (Makman2) [inline bears/rest/RSTcheckBear.py]: could you make a new commit out of this change? It's good to use `typed_list(str)` here, however it's a different issue :)

File: bears/rest/RSTcheckBear.py [hunk 3]
```diff
@@ -32,16 +33,32 @@ class RSTcheckBear:
...
+             to ignore custom roles.
         """
         args = ()
         if code_block_language_ignore:
             args = ('--ignore-language=' +
                     ','.join(code_block_language_ignore),)
-        return args + (filename,)
+        if directive_ignore:
+            directives = ','.join(part.strip() for part in directive_ignore)
+            args = ('--ignore-directives=' + directives,)
```
Comment (Makman2) [inline bears/rest/RSTcheckBear.py]: you overwrite the `args` if `code_block_language_ignore` was set.

Comment (Makman2) [reply bears/rest/RSTcheckBear.py]: you need to append to the `args`: `args += (...)`

I would suggest you to use a list, as appending to lists is usually faster than for tuples :+1:

File: bears/rest/RSTcheckBear.py [hunk 4]
```diff
@@ -32,16 +33,32 @@ class RSTcheckBear:
...
+        if role_ignore:
+            roles = ','.join(part.strip() for part in role_ignore)
+            args = ('--ignore-roles=' + roles,)
```
Comment (Makman2) [inline bears/rest/RSTcheckBear.py]: also here, you overwrite `args` if one of the settings before was set :+1:

File: bears/rest/RSTcheckBear.py [hunk 5]
```diff
@@ -32,16 +33,32 @@ class RSTcheckBear:
...
         """
         :param code_block_language_ignore:
             Comma seperated list for languages of which code blocks
             should be ignored. Code block of following languages
             can be detected: ``bash``, ``c``, ``cpp``, ``json``,
             ``python``, ``rst``.
+
+        :param directive_ignore:
+             Comma separated list of directives to be ignored. Can be
+             used to ignore custom directives as well.
```
Comment (Makman2) [inline bears/rest/RSTcheckBear.py]: "as well"? :)

**Comment by Makman2** [comment]
Could you squash all commits together (except the change that I mentioned belonging to an own commit)? :)
Please keep the `Closes XXX` reference in the commit body like @mridubhatnagar's commit already has :+1:

File: bears/rest/RSTcheckBear.py (hunk 5 above)
Comment (umeshksingla) [reply bears/rest/RSTcheckBear.py]: removed.

File: bears/rest/RSTcheckBear.py (hunk 1 above)
Comment (umeshksingla) [reply bears/rest/RSTcheckBear.py]: removed.

File: bears/rest/RSTcheckBear.py (hunk 2 above)
Comment (umeshksingla) [reply bears/rest/RSTcheckBear.py]: done.

File: bears/rest/RSTcheckBear.py (hunk 4 above)
Comment (umeshksingla) [reply bears/rest/RSTcheckBear.py]: oops

File: bears/rest/RSTcheckBear.py (hunk 3 above)
Comment (umeshksingla) [reply bears/rest/RSTcheckBear.py]: and then convert it back to tuple while returning?

File: bears/rest/RSTcheckBear.py [hunk 6]
```diff
@@ -32,16 +33,30 @@ class RSTcheckBear:
...
+             Comma separated list of roles to be ignored. Can be used
+             to ignore custom roles.
         """
         args = ()
         if code_block_language_ignore:
             args = ('--ignore-language=' +
                     ','.join(code_block_language_ignore),)
-        return args + (filename,)
+        if directive_ignore:
+            directives = ','.join(part.strip() for part in directive_ignore)
```
Comment (Makman2) [inline bears/rest/RSTcheckBear.py]: is stripping actually necessary?

Comment (umeshksingla) [reply bears/rest/RSTcheckBear.py]: It is necessary when you pass it to rstcheck, but here we don't know in what way user will pass the arguments

**Comment by Makman2** [comment]
Could you give some credit in the commit message to @mridubhatnagar ? Seems you have used at least a bit of her code :)

**Comment by umeshksingla** [comment]
I'll add "Thanks @mridubhatnagar for helping", works?
Also, CircleCI is failing every time? Do I need to check my tests?

**Comment by Makman2** [comment]
This is one of our dependencies, `language-check`. This happens sometimes. Try to rebase tomorrow and let's see what happens. If this persists we need to take a closer look. But this problem is very likely not caused by you :)

**Comment by mridubhatnagar** [comment]
Thanks @makman2 and @umeshksingla for taking this forward. I was stuck on the tests part becuase of which it was pending since long. 
Once, this gets approved. There is one of the issues blocked because of this. It can also be closed then.

**Comment by Makman2** [comment]
@mridubhatnagar which one actually?

**Comment by Makman2** [comment]
After `language-check` could be installed after some rebuilds, this needs fixing:
```
Project wide:
**** GitCommitBear [Section: commit | Severity: NORMAL] ****
!    ! Invalid full issue reference: #1723
```
Please provide a full link inside the commit message instead the short `#1723` variant.

**Comment by mridubhatnagar** [comment]
@Makman2 https://github.com/coala/coala/pull/4211.

**Comment by umeshksingla** [comment]
coala/coala#4211 looks like updating coala repository to use RSTcheckBear, is it? But either this shouldn't be blocking that. This PR is simply an enhancement.

File: tests/rest/RSTcheckBearTest.py [hunk 7]
```diff
@@ -17,3 +17,60 @@
...
+:RFC:`793`
+'''
+
+bad_role = '''
+Bar
+===
+
+:bad:`345`
+'''
+
```
Comment (mridubhatnagar) [inline tests/rest/RSTcheckBearTest.py]: Just to make the code more readable. We can good_directive1, good_role just after bad file and good file example. Than maybe have all the 3 tests together.
code_block_language_ignore
ignoredirective
ignorerole

**Comment by mridubhatnagar** [comment]
@umeshksingla  Yes. It is for updating the repo to use rstCheckbear. But adding rstcheckbear in coafile raised error related to directive and roles. Hence, for resolving that error. A seperate issue for roles and directives was raised.

File: tests/rest/RSTcheckBearTest.py (hunk 7 above)
Comment (umeshksingla) [reply tests/rest/RSTcheckBearTest.py]: I just thought I'll add my stuff at the end, but yes it is a good idea and if that's what other test files follow in general in coala.

File: bears/rest/RSTcheckBear.py (hunk 6 above)
Comment (Makman2) [reply bears/rest/RSTcheckBear.py]: maybe coala strips such whitespaces automatically away (also trailing ones behind a comma), we should test that before doing another strip (which would be just waste of execution time and if the user explicitly wants to set whitespaces using escapes - although unlikely here - we would properly fail)

Comment (Makman2) [reply bears/rest/RSTcheckBear.py]: Okay I tested it, you can safely leave the `strip()`. Whitespaces inside the lists you pass via a coafile are automatically stripped from each element, so these are the same for example:
```
directive_ignore = a,b      , c     ,     d, e
directive_ignore = a, b, c, d, e
```

File: bears/rest/RSTcheckBear.py [hunk 8]
```diff
@@ -32,16 +33,30 @@ class RSTcheckBear:
...
+            args += ('--ignore-directives=' + directives,)
+        if role_ignore:
+            roles = ','.join(part.strip() for part in role_ignore)
+            args += ('--ignore-roles=' + roles,)
+        return args + (filename, )
```
Comment (Makman2) [inline bears/rest/RSTcheckBear.py]: redundant space after comma :+1:

File: bears/rest/RSTcheckBear.py (hunk 6 above)
Comment (umeshksingla) [reply bears/rest/RSTcheckBear.py]: Okay, but we want is to have "a,b,c,d,e" and not "a, b, c, d, e". `rstcheck` doesn't allow spaces between the names of the directives. Also, I saw stripping done in many bear files.

Comment (umeshksingla) [reply bears/rest/RSTcheckBear.py]: ^^ doesn't strip() only remove the whitespaces at the end/beginning of a string? A user would not most likely have an escaped whitespace at the end of a string.

Comment (Makman2) [reply bears/rest/RSTcheckBear.py]: yes, however coala already does this for you, you don't need to strip again :+1: As you've specified a `typed_list(str)`

**Comment by jayvdb** [comment]
Please rebase.  This commit has a parent from May!

**Comment by umeshksingla** [comment]
@jayvdb I simply did: 
```
git fetch origin
git rebase origin/master
git pull --rebase
git push mycopy
```
Is it okay to show 112 commits?

**Comment by aviaryan** [comment]
I didn't ask to be assigned, did I? 😯

**Comment by umeshksingla** [comment]
@aviaryan I think I screwed while rebasing and the bot took all the labels and assignees from the commits merged before and pounded them over here :/

**Comment by Makman2** [comment]
wee looks like a bug in gitmate-bot :3

File: bears/rest/RSTcheckBear.py [hunk 9]
```diff
@@ -32,16 +33,30 @@ class RSTcheckBear:
...
+        :param role_ignore:
+             Comma separated list of roles to be ignored. Can be used
+             to ignore custom roles.
         """
         args = ()
         if code_block_language_ignore:
             args = ('--ignore-language=' +
                     ','.join(code_block_language_ignore),)
+        if directive_ignore:
+            directives = ','.join(part.strip() for part in directive_ignore)
```
Comment (Makman2) [inline bears/rest/RSTcheckBear.py]: like said, you don't need `part.strip()`, just do `','.join(directive_ignore)`.
You can even inline it into the next line:
```python
args += ('--ignore-directives=' + ','.join(directive_ignore),)
```

File: bears/rest/RSTcheckBear.py [hunk 10]
```diff
@@ -32,16 +33,30 @@ class RSTcheckBear:
...
+            args += ('--ignore-directives=' + directives,)
+        if role_ignore:
+            roles = ','.join(part.strip() for part in role_ignore)
```
Comment (Makman2) [inline bears/rest/RSTcheckBear.py]: see above

File: bears/rest/RSTcheckBear.py (hunk 9 above)
Comment (umeshksingla) [reply bears/rest/RSTcheckBear.py]: Sure.

**Comment by Makman2** [comment]
@umeshksingla what about https://github.com/coala/coala-bears/pull/2018/commits/20cb78faca1941f8d37d373eeaeb178f961c72f1#r135389951 ? :)

**Comment by Makman2** [comment]
Though I don't mind, imo it's okay like this. If you want to change it like @mridubhatnagar suggested or if you, @mridubhatnagar, want it changed, feel free to file an issue. Imo both styles are okay: A. mixing test data and tests or B. test data first and then tests :smile:

**Comment by Makman2** [comment]
ack 20cb78f

**Comment by umeshksingla** [comment]
@Makman2 I saw both kinds in the existing tests for bears, so..

**Comment by Makman2** [comment]
@gitmate-bot rebase

**Comment by Makman2** [comment]
@rultor merge

**Comment by umeshksingla** [comment]
@Makman2 Thanks :) how soon is this going to get included in your next release? :P

**Comment by Makman2** [comment]
Could take a while. We try to do releases quarterly, however you can already use the feature by installing the latest development-version: `pip3 install coala-bears --pre`

**Comment by umeshksingla** [comment]
Thanks again, @Makman2, though I use coala through tox.

**Pull request merged**